File: legacy/promotion.js
"use strict";

/**
 * promotion.js — Belief Formation / Long-Term Potentiation Layer
 *
 * Biology analog: Repeated, stable, goal-aligned co-activation with low
 * contradiction earns a pattern permanent synaptic change (LTP). Here,
 * a dream candidate that meets all thresholds gets written back into
 * universe as durable memory — it becomes a belief.
 *
 * Promotion requires ALL of:
 *   1. seenCount >= THRESHOLDS.seenCount        (must recur, not one-shot)
 *   2. bestC >= THRESHOLDS.bestC                (commit readiness proven)
 *   3. bestPhi_g >= THRESHOLDS.bestPhi_g        (emergence quality proven)
 *   4. bestConfidence >= THRESHOLDS.bestConf    (cleanup match quality)
 *   5. meanContradiction <= THRESHOLDS.maxChi   (contradiction must be low)
 *   6. contradictionStddev <= THRESHOLDS.maxChiSd  (must be STABLE, not spiking)
 *   7. nonSourceRatio >= THRESHOLDS.minNSR      (must be genuine insight, not echo)
 *
 * Competition: before promotion, candidates within the same "cluster"
 * (resonance > 0.72 between their universe vectors) compete — only the
 * highest-scoring wins; others are suppressed for this cycle.
 */

const universe        = require('./universe');
const candidateBuffer = require('./candidate-buffer');
const { textVec, resonance } = require('./rshl-core');
const { clamp01, mean, stddev } = require('./field-state');

// ── Promotion thresholds ──────────────────────────────────────────────────────
// Calibrated to the actual field output range.
// With rho ≈ activeCount/totalCount (e.g. 2/33 ≈ 0.09 for a seeded universe),
// phi_g peaks around 0.03–0.06 and C peaks around 0.015–0.040.
// Thresholds must be set relative to what the field can actually produce —
// not abstract [0,1] ideals — so that promotion is achievable but not trivial.
const THRESHOLDS = {
    seenCount:   3,     // Must recur in at least 3 independent dream cycles
    bestC:       0.015, // Minimum commit readiness (C = phi_g × (1-chi) × tau)
    bestPhi_g:   0.024, // Minimum integrated goal-aligned emergence
    bestConf:    0.72,  // Minimum attractor cleanup confidence (field scores high here)
    maxChi:      0.38,  // Mean contradiction must stay below this
    maxChiSd:    0.28,  // Contradiction must be stable (not oscillating)
    minNSR:      0.35,  // At least 35% of sightings must be non-source insight
    competeSim:  0.72,  // Vector similarity threshold for competition grouping
};

// ── Scoring ───────────────────────────────────────────────────────────────────
function _score(entry) {
    const nsr       = entry.seenCount > 0 ? entry.nonSourceCount / entry.seenCount : 0;
    const chiMean   = mean(entry.contradictionHistory);
    const stability = clamp01(1 - stddev(entry.phiHistory || [entry.bestPhi_g]));
    return (
        entry.bestPhi_g   * 0.30 +
        entry.bestC       * 0.25 +
        entry.bestConf    * 0.15 +
        nsr               * 0.15 +
        stability         * 0.10 +
        clamp01(1 - chiMean) * 0.05
    );
}

// ── Gate check ────────────────────────────────────────────────────────────────
function _passesThresholds(entry) {
    if (entry.status !== candidateBuffer.STATUS.CANDIDATE) return { pass: false, reason: 'not-candidate' };
    if (entry.seenCount < THRESHOLDS.seenCount)            return { pass: false, reason: 'seen-count' };
    if (entry.bestC     < THRESHOLDS.bestC)                return { pass: false, reason: 'best-C' };
    if (entry.bestPhi_g < THRESHOLDS.bestPhi_g)            return { pass: false, reason: 'best-phi_g' };
    if (entry.bestConfidence < THRESHOLDS.bestConf)        return { pass: false, reason: 'best-confidence' };

    const ch = entry.contradictionHistory;
    if (!ch || !ch.length)                                 return { pass: false, reason: 'no-chi-history' };

    const chiMean = mean(ch);
    if (chiMean > THRESHOLDS.maxChi)                       return { pass: false, reason: 'chi-too-high' };

    const chiSd = stddev(ch);
    if (chiSd > THRESHOLDS.maxChiSd)                       return { pass: false, reason: 'chi-unstable' };

    const nsr = entry.seenCount > 0 ? entry.nonSourceCount / entry.seenCount : 0;
    if (nsr < THRESHOLDS.minNSR)                           return { pass: false, reason: 'echo-ratio' };

    return { pass: true };
}

// ── Competition: cluster threshold-passing candidates by vector similarity ────
// Within each cluster, only the highest scorer is allowed to promote this cycle.
function _resolveCompetition(eligibles) {
    if (!eligibles.length) return [];

    // Compute vectors for each eligible candidate
    const withVecs = eligibles.map(e => ({
        entry: e,
        vec: textVec(e.text),
        score: _score(e),
    }));

    const suppressed = new Set();
    const winners    = [];

    for (let i = 0; i < withVecs.length; i++) {
        if (suppressed.has(i)) continue;
        let bestIdx = i;

        for (let j = i + 1; j < withVecs.length; j++) {
            if (suppressed.has(j)) continue;
            const sim = clamp01(resonance(withVecs[i].vec, withVecs[j].vec));
            if (sim >= THRESHOLDS.competeSim) {
                // Same cluster — keep the higher scorer, suppress the other
                if (withVecs[j].score > withVecs[bestIdx].score) {
                    suppressed.add(bestIdx);
                    bestIdx = j;
                } else {
                    suppressed.add(j);
                }
            }
        }

        if (!suppressed.has(bestIdx)) {
            winners.push(withVecs[bestIdx].entry);
        }
    }

    return winners;
}

// ── Main promotion run ────────────────────────────────────────────────────────
function runPromotion() {
    const all       = candidateBuffer.getAll();
    const eligible  = [];
    const failLog   = [];

    for (const entry of all) {
        const check = _passesThresholds(entry);
        if (check.pass) {
            eligible.push(entry);
        } else if (entry.status === candidateBuffer.STATUS.CANDIDATE) {
            failLog.push({ key: entry.key, reason: check.reason });
        }
    }

    const winners   = _resolveCompetition(eligible);
    const promoted  = [];

    for (const entry of winners) {
        const chiMean   = mean(entry.contradictionHistory);
        const stability = clamp01(1 - stddev(entry.phiHistory || [entry.bestPhi_g]));

        // Map field quality into universe strength (range 1.5–4.0)
        const rawStrength = clamp01(entry.bestC * 2.5 + entry.bestPhi_g * 1.5 + stability * 0.5);
        const strength    = 1.5 + rawStrength * 2.5; // maps [0,1] → [1.5, 4.0]

        universe.store(entry.text, 'memory', {
            source:        'promoted-dream',
            strength,
            novelty:       clamp01(entry.bestPhi_g),
            contradiction: chiMean,
            unresolved:    false,
            promotedAt:    Date.now(),
            seenCount:     entry.seenCount,
            bestC:         entry.bestC,
            bestPhi_g:     entry.bestPhi_g,
            score:         _score(entry),
        });

        candidateBuffer.markPromoted(entry.key);
        promoted.push({
            text:     entry.text,
            seenCount: entry.seenCount,
            bestC:    entry.bestC,
            bestPhi_g: entry.bestPhi_g,
            strength,
        });
    }

    return { promoted, failLog, eligible: eligible.length };
}

module.exports = {
    THRESHOLDS,
    runPromotion,
};

File: kai-rust/src/cognition/promotion.rs
/// Promotion — Belief formation / Long-term potentiation.
///
/// A dream candidate that meets ALL thresholds gets written into
/// the universe as durable memory — it becomes a belief.
use super::candidates::{CandidateBuffer, CandidateStatus};
use crate::core::Universe;

pub struct PromotionThresholds {
    pub seen_count: u32,
    pub best_c: f32,
    pub best_phi_g: f32,
    pub best_confidence: f32,
    pub max_chi: f32,
    pub max_chi_sd: f32,
    pub min_nsr: f32,
}

impl Default for PromotionThresholds {
    fn default() -> Self {
        Self {
            seen_count: 3,
            best_c: 0.015,
            best_phi_g: 0.024,
            best_confidence: 0.72,
            max_chi: 0.38,
            max_chi_sd: 0.28,
            min_nsr: 0.35,
        }
    }
}

#[derive(Debug)]
pub struct PromotionResult {
    pub promoted: Vec<PromotedBelief>,
    pub eligible: usize,
}

#[derive(Debug)]
pub struct PromotedBelief {
    pub text: String,
    pub seen_count: u32,
    pub best_c: f32,
    pub best_phi_g: f32,
    pub strength: f32,
}

fn mean(v: &[f32]) -> f32 {
    if v.is_empty() {
        return 0.0;
    }
    v.iter().sum::<f32>() / v.len() as f32
}

fn stddev(v: &[f32]) -> f32 {
    if v.len() < 2 {
        return 0.0;
    }
    let m = mean(v);
    let var = v.iter().map(|x| (x - m).powi(2)).sum::<f32>() / v.len() as f32;
    var.sqrt()
}

#[allow(dead_code)]
fn score(e: &super::candidates::CandidateEntry) -> f32 {
    let nsr = if e.seen_count > 0 {
        e.non_source_count as f32 / e.seen_count as f32
    } else {
        0.0
    };
    let chi_mean = mean(&e.contradiction_history);
    let stability = (1.0 - stddev(&e.phi_history)).clamp(0.0, 1.0);
    e.best_phi_g * 0.30
        + e.best_c * 0.25
        + e.best_confidence * 0.15
        + nsr * 0.15
        + stability * 0.10
        + (1.0 - chi_mean).clamp(0.0, 1.0) * 0.05
}

/// Run promotion check on all candidates.
pub fn run_promotion(
    candidates: &mut CandidateBuffer,
    universe: &mut Universe,
    thresholds: &PromotionThresholds,
) -> PromotionResult {
    let all = candidates.get_all();
    let mut eligible_keys: Vec<String> = Vec::new();

    for entry in &all {
        if entry.status != CandidateStatus::Candidate {
            continue;
        }
        if entry.seen_count < thresholds.seen_count {
            continue;
        }
        if entry.best_c < thresholds.best_c {
            continue;
        }
        if entry.best_phi_g < thresholds.best_phi_g {
            continue;
        }
        if entry.best_confidence < thresholds.best_confidence {
            continue;
        }

        let chi_mean = mean(&entry.contradiction_history);
        if chi_mean > thresholds.max_chi {
            continue;
        }

        let chi_sd = stddev(&entry.contradiction_history);
        if chi_sd > thresholds.max_chi_sd {
            continue;
        }

        let nsr = if entry.seen_count > 0 {
            entry.non_source_count as f32 / entry.seen_count as f32
        } else {
            0.0
        };
        if nsr < thresholds.min_nsr {
            continue;
        }

        eligible_keys.push(entry.key.clone());
    }

    let eligible_count = eligible_keys.len();
    let mut promoted = Vec::new();

    // Competition: among eligible, resolve by vector similarity clusters
    // For now, promote all that pass (competition requires vector comparison)
    for key in &eligible_keys {
        let entry = candidates.get_all().into_iter().find(|e| e.key == *key);
        if let Some(entry) = entry {
            let stability = (1.0 - stddev(&entry.phi_history)).clamp(0.0, 1.0);
            let raw_strength =
                (entry.best_c * 2.5 + entry.best_phi_g * 1.5 + stability * 0.5).clamp(0.0, 1.0);
            let strength = 1.5 + raw_strength * 2.5; // maps [0,1] → [1.5, 4.0]

            universe.store(&entry.text, "memory", "promoted-dream", strength);

            promoted.push(PromotedBelief {
                text: entry.text.clone(),
                seen_count: entry.seen_count,
                best_c: entry.best_c,
                best_phi_g: entry.best_phi_g,
                strength,
            });
        }
    }

    for key in &eligible_keys {
        candidates.mark_promoted(key);
    }

    PromotionResult {
        promoted,
        eligible: eligible_count,
    }
}
